OpenRA.Mods.Common/Traits/Armament.cs

```diff
@@ -253,24 +261,28 @@ protected virtual bool CanFire(Actor self, in Target target)
 
 		// Note: facing is only used by the legacy positioning code
 		// The world coordinate model uses Actor.Orientation
+		Barrel barrel;
 		public virtual Barrel CheckFire(Actor self, IFacing facing, in Target target)
 		{
 			if (!CanFire(self, target))
 				return null;
-
 			if (ticksSinceLastShot >= Weapon.ReloadDelay)
 				Burst = Weapon.Burst;
 
 			ticksSinceLastShot = 0;
 
-			// If Weapon.Burst == 1, cycle through all LocalOffsets, otherwise use the offset corresponding to current Burst
-			currentBarrel %= barrelCount;
-			var barrel = Weapon.Burst == 1 ? Barrels[currentBarrel] : Barrels[Burst % Barrels.Length];
-			currentBarrel++;
+			for (var i = 0; i < Info.BurstsPerFire && Burst > 0; i++)
+			{
+				// If Weapon.Burst == 1, cycle through all LocalOffsets, otherwise use the offset corresponding to current Burst
+				currentBarrel %= barrelCount;
+				barrel = Weapon.Burst == 1 ? Barrels[currentBarrel] : Barrels[Burst % Barrels.Length];
+				currentBarrel++;
 
-			FireBarrel(self, facing, target, barrel);
+				FireBarrel(self, facing, target, barrel);
+				Burst--;
```

Comment:
Bursts for all barrels will now always have to run in sync? No use case anticipated where they should vary?

Reply:
Actually when BurstsPerFire > 0, the barrels in the same BurstsPerFire loop is in sync.

If BurstsPerFire default == 1, the barrels will work like its used to be. 

The use case is Juggernaut in the TS, you can see its barrels is not in sync when fire